testsuite/devserver.go

```diff
@@ -0,0 +1,356 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package testsuite contains unit testing framework for Temporal workflows and activities and a helper to download and
+// start a dev server.
+package testsuite
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"syscall"
+	"time"
+
+	"go.temporal.io/sdk/client"
+	"go.temporal.io/sdk/internal"
+	ilog "go.temporal.io/sdk/internal/log"
+	"go.temporal.io/sdk/log"
+)
+
+// Cached download of the dev server.
+type CachedDownload struct {
+	// Which version to download, by default the latest version compatible with the SDK will be downloaded.
+	// Acceptable values are specific release versions (e.g v0.3.0), "default", and "latest".
+	Version string
+	// Destination directory or the user temp directory if unset.
+	DestDir string
+}
+
+// Where to find an executable. Can be a path or cached download.
+type EphemeralExe struct {
+	// Existing path on the filesystem for the executable.
+	ExistingPath string
+	// Download the executable if not already there.
+	CachedDownload CachedDownload
+}
+
+// Configuration for the dev server.
+type DevServerOptions struct {
+	// Path to executable or download info - defaults to cached download.
+	Exe EphemeralExe
+	// Client options used to create a client for the dev server.
+	// The provided Namespace or the "default" namespace is automatically registered on startup.
+	// If HostPort is provided, the host and port will be used to bind the server, otherwise the server will bind to
+	// localhost and obtain a free port.
+	ClientOptions *client.Options
+	// SQLite DB filename if persisting or non-persistent if none.
+	DBFilename string
+	// Whether to enable the UI.
+	EnableUI bool
+	// Log format - defaults to "pretty"
+	LogFormat string
+	// Log level - defaults to "warn"
+	LogLevel string
+	// Additional arguments to the dev server.
+	ExtraArgs []string
+}
+
+type DevServer interface {
```

Comment:
Why an interface instead of just making the `devServer` exported? If this is potential reuse for test server, we need a better interface name, because `DevServerOptions` also uses "dev server" term but the args clearly wouldn't work for test server.

Reply:
We agreed to return an interface to support `temporaltest`, Java test server, and CLI.

I might rename it though, just don't have a better name.